File: backend/Signa.Tests.Common/Builders/Domain/Features/FingerBuilder.cs
using Signa.Domain.Features;

namespace Signa.Tests.Common.Builders.Domain.Features
{
    public class FingerBuilder
    {
        private double[] tipDirection;
        private FingerType type;

        public FingerBuilder WithDirection(double[] tipDirection)
        {
            this.tipDirection = tipDirection;
            return this;
        }

        public FingerBuilder OfType(FingerType type)
        {
            this.type = type;
            return this;
        }

        public Finger Build()
        {
            return new Finger
            {
                TipDirection = tipDirection,
                Type = type
            };
        }

        public static Finger Thumb()
        {
            return new FingerBuilder()
                .WithDirection(TipDirection())
                .OfType(FingerType.Thumb)
                .Build();
        }

        public static Finger Index()
        {
            return new FingerBuilder()
                .WithDirection(TipDirection())
                .OfType(FingerType.Index)
                .Build();
        }

        public static Finger Middle()
        {
            return new FingerBuilder()
                .WithDirection(TipDirection())
                .OfType(FingerType.Middle)
                .Build();
        }

        public static Finger Ring()
        {
            return new FingerBuilder()
                .WithDirection(TipDirection())
                .OfType(FingerType.Ring)
                .Build();
        }

        public static Finger Pinky()
        {
            return new FingerBuilder()
                .WithDirection(TipDirection())
                .OfType(FingerType.Pinky)
                .Build();
        }

        private static double[] TipDirection()
        {
            return new DoubleArrayBuilder().WithSize(3).Build();
        }

        public static Finger[] DefaultFingers()
        {
            return new[] 
            { 
                FingerBuilder.Thumb(), 
                FingerBuilder.Index(), 
                FingerBuilder.Middle(), 
                FingerBuilder.Ring(), 
                FingerBuilder.Pinky() 
            };
        }
    }
}

File: frontend/js/tests/builders/FingerBuilder.js
function FingerBuilder(){}
FingerBuilder.FingerType = {
    THUMB: 0,
    INDEX: 1,
    MIDDLE: 2,
    RING: 3,
    PINKY: 4
};

FingerBuilder.prototype = {
    _direction: undefined,
    _type: FingerBuilder.FingerType.THUMB,

    withDirection: function(direction)
    {
        this._direction = direction;
        return this;
    },

    ofType: function(type)
    {
        this._type = type;
        return this;
    },

    build: function()
    {
        return {
            direction: this._direction,
            type: this._type
        }
    }
};

function _fingerOfType(type)
{
    var direction = Math.random() * type; // some magical random position
    return new FingerBuilder()
        .withDirection([direction, direction, direction])
        .ofType(type)
        .build();
}

FingerBuilder.thumb = function() { return _fingerOfType(FingerBuilder.FingerType.THUMB); };
FingerBuilder.index = function() { return _fingerOfType(FingerBuilder.FingerType.INDEX); };
FingerBuilder.middle = function() { return _fingerOfType(FingerBuilder.FingerType.MIDDLE); };
FingerBuilder.ring = function() { return _fingerOfType(FingerBuilder.FingerType.RING); };
FingerBuilder.pinky = function() { return _fingerOfType(FingerBuilder.FingerType.PINKY); };

module.exports = FingerBuilder;
